libraries/AP_Soaring/SoaringController.h

```diff
@@ -0,0 +1,167 @@
+/* 
+Soaring Controller class by Samuel Tabor
+Provides a layer between the thermal centring algorithm and the main code for managing navigation targets, data logging, tuning parameters, algorithm inputs and eventually other soaring strategies such as speed-to-fly. AP_TECS libary used for reference.
+*/
+#ifndef SoaringController_h
+#define SoaringController_h
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_Param/AP_Param.h>
+#include <DataFlash/DataFlash.h>
+#include "MatrixMath.h"
+#include <AP_Math/AP_Math.h>
+#include "ExtendedKalmanFilter.h"
+#include <AP_SpdHgtControl/AP_SpdHgtControl.h>
+
+#define EXPECTED_THERMALLING_SINK 0.7
+#define INITIAL_THERMAL_STRENGTH 2.0
+#define INITIAL_THERMAL_RADIUS 30.0 //150.0
+#define INITIAL_STRENGTH_COVARIANCE 0.0049
+#define INITIAL_RADIUS_COVARIANCE 2500.0
+#define INITIAL_POSITION_COVARIANCE 300.0
+#define ASPD_FILT 0.05
+#define TE_FILT 0.03
+#define TE_FILT_DISPLAYED 0.15
+#define EKF_MAX_BUFFER_SIZE 300
+
+class SoaringController
+{
+    ExtendedKalmanFilter _ekf;
+    AP_AHRS &_ahrs;
+    AP_SpdHgtControl *&_spdHgt;
+    const AP_Vehicle::FixedWing &_aparm;
+    const float rate_hz = 5;
+  
+    // Keep track of the waypoint so we can restore after coming out of thermal mode.
+    struct Location _prev_next_wp;
+
+    //Store aircraft location at last update
+    struct Location _prev_update_location;
+
+    //Store aircraft location at last update
+    struct Location _prev_vario_update_location;
+
+    // store time thermal was entered for hysteresis
+    unsigned long _thermal_start_time_us;
+
+    // store time cruise was entered for hysteresis
+    unsigned long _cruise_start_time_us;
+
+    // store time of last update
+    unsigned long _prev_update_time;
+
+    // store time of last update
+    unsigned long _prev_vario_update_time;
+    //gcs_send_text_P(SEVERITY_LOW, PSTR("Soar initialisation complete"));
+
+    float _vario_reading;
+    float _filtered_vario_reading;
+    float _last_alt;
+    float _alt;
+    float _last_aspd;
+    float _last_roll;
+    float _last_total_E;
+    bool _new_data;
+    float _loiter_rad; // Loiter radius passed in
+    bool _throttle_suppressed;
+
+    float _ekf_buffer[EKF_MAX_BUFFER_SIZE][3];
+    unsigned _ptr;
+    unsigned _nsamples;
+    Location dummy = Location();
```

Comment:
Because a reference variable can't be set to the return location of a method call. The idea of a  reference variable in C++ is that it enables you to modify whatever is stored in the memory location that it references. So, you can use a reference variable to modify the value of another variable (e.g., dummy), but you can't use to modify certain special locations such as where the return value of a method call is stored.
Because a reference variable can't be set to the return location of a method call. The idea of a  reference variable in C++ is that it enables you to modify whatever is stored in the memory location that it references. So, you can use a reference variable to modify the value of another variable (e.g., dummy), but you can't use to modify values stored in certain special locations such as where the return value of a method call is stored.